### PR TITLE
Split function in sc.get

### PR DESCRIPTION
**Pull request by Koncopd**
This adds `sc.get.split` to split adata based on an obs column. The examples are in the docstring.

### CODE REVIEW

**Comment by ivirshup** [comment]
I think this is a really good idea. Definitely a good hold over until a real GroupBy object is worked out. I'll review in depth later but have a few thoughts:

* How about `split_by`? Rhymes with `group_by` which this is quite like.
* I would like to be able to do this along the `var` axis
* What do you think about including some of the grouping options from [`_prepare_dataframe`](https://github.com/theislab/scanpy/blob/d072abd05bda07f280ea91f5e7e4a84f9782c118/scanpy/plotting/_anndata.py#L1842)? E.g. grouping by multiple keys, binning by continuous values? It would be nice if this could be used in place of `_prepare_dataframe`.

The features would not have to be implemented in this PR, but it would be good to think about what the API for them would be and making sure it could be added elegantly.

**Comment by Koncopd** [comment]
Renamed to `split_by`, added `axis` argument.

**Comment by Koncopd** [comment]
What should be names of interval categories in a returned `dict` for binning? Intervals converted to strings?

File: scanpy/get/get.py [hunk 1]
```diff
@@ -437,3 +437,141 @@ def _set_obs_rep(adata, val, *, use_raw=False, layer=None, obsm=None, obsp=None)
...
+    if axis not in (0, 1):
+        raise ValueError("axis should be 0 or 1 only.")
+
+    attr_by_key = adata.obs[key] if axis == 0 else adata.var[key]
+    select = [slice(None), slice(None)]
+
+    adatas = {}
+
+    if groups is None:
+        groups = np.unique(attr_by_key)
```
Comment (ivirshup) [inline scanpy/get/get.py]: This should be faster for categoricals (since pandas knows to just look at the categories)

```suggestion
        groups = attr_by_key.unique()
```

File: scanpy/get/get.py [hunk 2]
```diff
@@ -437,3 +437,141 @@ def _set_obs_rep(adata, val, *, use_raw=False, layer=None, obsm=None, obsp=None)
...
+                                 dict(some=['CD14+ Monocyte', 'CD34+']),
+                                 others_key='others')
+    >>> adatas
+    {'some': View of AnnData object with n_obs × n_vars = 142 × 765
+     ...,
+     'others': View of AnnData object with n_obs × n_vars = 558 × 765
+     ...
+    }
+    """
+    if key not in adata.obs:
```
Comment (ivirshup) [inline scanpy/get/get.py]: This wouldn't work for the var axis. Maybe you could abstract over the dimension before this check?

File: scanpy/get/get.py [hunk 3]
```diff
@@ -437,3 +437,141 @@ def _set_obs_rep(adata, val, *, use_raw=False, layer=None, obsm=None, obsp=None)
...
+    copy: bool = False,
+) -> Dict[str, AnnData]:
+    """\
+    Split adata by obs key.
+
+    Params
+    ------
+    adata
+        AnnData object to split.
+    key
```
Comment (ivirshup) [inline scanpy/get/get.py]: How about calling this `by` to mimic `pd.DataFrame.groupby` or `group` for `xarray.groupby`?

File: scanpy/get/get.py [hunk 4]
```diff
@@ -437,3 +437,141 @@ def _set_obs_rep(adata, val, *, use_raw=False, layer=None, obsm=None, obsp=None)
...
+
+GroupsList = List[str]
+GroupsLists = List[GroupsList]
+GroupsDict = Dict[str, GroupsList]
+
+
+def split_by(
+    adata: AnnData,
+    key: str,
+    groups: Optional[Union[GroupsList, GroupsLists, GroupsDict]] = None,
```
Comment (ivirshup) [inline scanpy/get/get.py]:
```suggestion
    key: str,
    *.
    groups: Optional[Union[GroupsList, GroupsLists, GroupsDict]] = None,
```

Could we everything past `key` keyword only? This way we don't have to worry about changing orders.

**Review by ivirshup — COMMENTED** [review]
I've taken a closer look. The argument I have the most issue with is `groups`, which I think could be replicated by allowing passing `Series` through for `key` (like `df.groupby(series)`).

# Specifying groups more granularly

I see the value in `GroupsList` and `GroupsDict`, but less `GroupsLists`. `GroupsLists` is just not providing a name for `GroupsDict`. I think I'd rather the user just had to specify a name for the grouping.

## `groups` argument usage and precedence

Also, can you provide precedence in other APIs for this kind of argument? I'm wondering if we should instead just allow passing a series as `key`, then the user can create whatever kind of flat grouping they want (example below with numeric values).

I'm also wondering if there is a reason other libraries (`xarray`, `pandas`) don't include this, and if defining the group sets/ subsetting the data makes more sense as a separate operation. Being able to regroup by `dict` seems quite elegant. Filtering the groups used seems like it's largely equivalent to subsetting.

## Non-flat labellings

One interesting part about providing sets of groupings is that they are no longer mutually exclusive sets. In scikit-learn this is represented by a 2d boolean array with dimensions `(obs x groups)` ([sklearn docs](https://scikit-learn.org/stable/modules/multiclass.html#id5)). It would be nice to have interchange with that format.

# Numeric values

> What should be names of interval categories in a returned dict for binning? Intervals converted to strings?

I think using the intervals themselves should be fine.

I'm wondering if we should even directly include this, or make this available through:

```python
sc.get.split_by(adata, pd.cut(adata.obs["total_counts"], bins=10))
```

I think this kind of solution also provides a way for customizing group selections (e.g. could replace the need for the `groups` argument).

# Library choice

Should this go in `anndata`?

File: scanpy/get/get.py (hunk 3 above)
Comment (Koncopd) [reply scanpy/get/get.py]: Renamed.

File: scanpy/get/get.py (hunk 2 above)
Comment (Koncopd) [reply scanpy/get/get.py]: Fixed

File: scanpy/get/get.py (hunk 1 above)
Comment (Koncopd) [reply scanpy/get/get.py]: Changed as you suggested.

**Comment by Koncopd** [comment]
Now `key` is renamed to `by`, it can take a string, list or series with group values for each observation. I still prefer to have the `groups` argument instead of making custom grouping series.
